File: src/main/java/org/lanternpowered/server/scheduler/LanternScheduledTask.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.scheduler;

import org.lanternpowered.api.util.ToStringHelper;
import org.spongepowered.api.scheduler.ScheduledTask;
import org.spongepowered.api.scheduler.Task;

import java.util.UUID;
import java.util.concurrent.Future;

import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * An internal representation of a {@link Task} created by a plugin.
 */
public class LanternScheduledTask implements ScheduledTask {

    private final UUID uniqueId;
    private final String name;
    final LanternTask task;
    private final LanternScheduler scheduler;

    @Nullable private Future<?> future;
    private final Object futureLock = new Object();

    volatile boolean scheduledRemoval;

    LanternScheduledTask(Task task, LanternScheduler scheduler) {
        this.scheduler = scheduler;
        this.uniqueId = UUID.randomUUID();
        this.task = (LanternTask) task;
        this.name = task.getName() + "-" + this.task.scheduledCounter.incrementAndGet();
    }

    @Override
    public UUID getUniqueId() {
        return this.uniqueId;
    }

    @Override
    public String getName() {
        return this.name;
    }

    @Override
    public Task getTask() {
        return this.task;
    }

    @Override
    public boolean cancel() {
        return cancel(false);
    }

    public boolean cancel(boolean mayInterruptIfRunning) {
        final Future<?> future = getFuture();
        if (future.isDone()) {
            return false;
        }
        // If the task is successfully cancelled, just
        // remove it from the scheduler
        if (future.cancel(mayInterruptIfRunning)) {
            this.scheduler.remove(this);
        } else {
            this.scheduledRemoval = true;
        }
        return false;
    }

    void setFuture(Future<?> future) {
        synchronized (this.futureLock) {
            this.future = future;
            this.futureLock.notifyAll();
        }
    }

    Future<?> getFuture() {
        synchronized (this.futureLock) {
            while (this.future == null) {
                try {
                    this.futureLock.wait();
                } catch (InterruptedException ignored) {
                }
            }
        }
        return this.future;
    }

    @Override
    public String toString() {
        return new ToStringHelper("ScheduledTask")
                .add("uniqueId", this.uniqueId)
                .add("name", this.name)
                .add("task", this.task)
                .toString();
    }
}

File: src/main/kotlin/org/lanternpowered/server/scheduler/LanternScheduledTask.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.scheduler

import org.lanternpowered.api.util.ToStringHelper
import org.spongepowered.api.scheduler.ScheduledTask
import org.spongepowered.api.scheduler.Task
import java.util.UUID
import java.util.concurrent.Future

/**
 * An internal representation of a [Task] created by a plugin.
 */
class LanternScheduledTask internal constructor(
        private val task: LanternTask,
        private val scheduler: LanternScheduler
) : ScheduledTask {

    private val uniqueId: UUID = UUID.randomUUID()
    private val name: String = this.task.name + "-" + this.task.scheduledCounter.incrementAndGet()
    private val futureLock = Any()
    @Volatile var scheduledRemoval = false

    var future: Future<*>? = null
        get() {
            synchronized(this.futureLock) {
                while (field == null) {
                    try {
                        @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
                        (this.futureLock as Object).wait()
                    } catch (ignored: InterruptedException) {
                    }
                }
            }
            return field
        }
        set(future) {
            synchronized(this.futureLock) {
                field = future
                @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
                (this.futureLock as Object).notifyAll()
            }
        }

    override fun getUniqueId(): UUID = this.uniqueId
    override fun getName(): String = this.name
    override fun getTask(): LanternTask = this.task

    override fun cancel(): Boolean = cancel(false)
    override fun isCancelled(): Boolean = this.future!!.isCancelled

    fun cancel(mayInterruptIfRunning: Boolean): Boolean {
        val future = this.future!!
        if (future.isDone)
            return false
        // If the task is successfully cancelled, just
        // remove it from the scheduler
        if (future.cancel(mayInterruptIfRunning)) {
            this.scheduler.remove(this)
            return true
        } else {
            this.scheduledRemoval = true
        }
        return false
    }

    override fun toString(): String = ToStringHelper("ScheduledTask")
            .add("uniqueId", this.uniqueId)
            .add("name", this.name)
            .add("task", this.task)
            .toString()
}
